src/client/components/forms/userCollection.js

```diff
@@ -0,0 +1,306 @@
+ * with this program; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+
+import * as bootstrap from 'react-bootstrap';
+import CustomInput from '../../input';
+import EntitySearchFieldOption from '../../entity-editor/common/entity-search-field-option';
+import PropTypes from 'prop-types';
+import React from 'react';
+import ReactSelect from 'react-select';
+import SelectWrapper from '../input/select-wrapper';
+import classNames from 'classnames';
+import request from 'superagent-bluebird-promise';
+
+
+const {Alert, Button, Col, Grid, Row} = bootstrap;
+
+class UserCollectionForm extends React.Component {
+	constructor(props) {
+		super(props);
+
+		this.state = {
+			collection: props.collection,
+			errorText: null
+		};
+
+		// React does not autobind non-React class methods
+		this.getCleanedCollaborators = this.getCleanedCollaborators.bind(this);
+		this.handleAddCollaborator = this.handleAddCollaborator.bind(this);
+		this.handleRemoveCollaborator = this.handleRemoveCollaborator.bind(this);
+		this.handleChangeCollaborator = this.handleChangeCollaborator.bind(this);
+		this.handleSubmit = this.handleSubmit.bind(this);
+		this.isValid = this.isValid.bind(this);
+	}
+
+	handleSubmit(evt) {
+		evt.preventDefault();
+
+		if (!this.isValid()) {
+			this.setState(prevState => ({
+				collection: prevState.collection,
+				errorText: 'Incomplete Form'
+			}));
+			return;
+		}
+
+		const collaborators = this.getCleanedCollaborators();
+		const description = this.description.getValue();
+		const name = this.name.getValue();
+		const privacy = this.privacy.getValue();
+		const entityType = this.entityType.getValue();
+
+		const data = {
+			collaborators,
+			description,
+			entityType,
+			name,
+			privacy
+		};
+
+		let submissionURL;
+		if (this.props.collection.id) {
+			submissionURL = `/collection/${this.props.collection.id}/edit/handler`;
+		}
+		else {
+			submissionURL = '/collection/create/handler';
+		}
+		request.post(submissionURL)
+			.send(data)
+			.promise()
+			.then((res) => {
+				if (res.status === 200) {
+					window.location.href = `/collection/${res.body.id}`;
+				}
+				else {
+					this.setState(prevState => ({
+						collection: prevState.collection,
+						errorText: `Internal Error: ${res.body.error}`
+					}));
+				}
+			});
+	}
+
+	isValid() {
+		return this.name.getValue() && this.entityType.getValue();
+	}
+
+	getCleanedCollaborators() {
+		const cleanedCollaborators = this.state.collection.collaborators.filter(collaborator => collaborator && collaborator.id !== null);
+		return cleanedCollaborators;
+	}
+
+	handleAddCollaborator() {
+		const newCollaborators = this.state.collection.collaborators;
+		newCollaborators.push({
+			id: null,
+			name: ''
+		});
+		this.setState(prevState => ({
+			collection: {
+				...prevState.collection,
+				collaborators: newCollaborators
```

Comment:
Hmm yess. I'll make this change

Reply:
I wonder if it wouldn't be even better to have collaborators separate on the state instead of a property of collection, so that React can optimize rendering better: if collaborators change but not the collection items, no need to redraw whatever uses the collection items. Currently I think the whole collection object in the state gets invalidated in this call.

It will require a bit of destructuring in the constructor.

Reply:
or this can be done too 
```
<UserCollectionForm
	canEditType={canEditType}
	collection={props.collection}
        collaborator={props.collaborator}
/>